### PR TITLE
Issue 1178/parallel events on map

### LINKED ISSUE #1178: Displaying parallel events in location modal 🌍
## Description
When a user is creating an event and picking a location for that event, they should be able to see what other locations have events happening on the same day.

## Relevant Job Stories
* When I select a location for an event I want to see what else is planned for this location, so that I can find the best fit for my event and avoid collisions

## Prerequisites
Don't start this until #1172 is done.

## Requirements
- [x] On the map, all pins should have space for a number
- [x] That number should indicate the number of events that take place on the relevant date
- [x] The relevant date is defined as the date of the event for which the modal is used to pick a location
- [x] If there are more than 9 events on the day, display 9+
- [x] If there are no events, display nothing (just the blank pin)
- [x] The card with details about selected location should display info about each event that takes place at this location on the relevant date
    - [x] If no relevant events, show icon
    - [x] Title of event should link to this event
    - [x] Chip should contain numbers of participants / number of required participants

## Possible implementations
Add a property to the modal to define the relevant date, and retrieve all events on that date, plotting them at their locations.

## Design specifications
image

Image

Image

## Open questions
None

### PR DESCRIPTION
**Pull request by ziggabyte**
## Description
This PR adds the number of events occurring at the same time to the markers on the map, and shows info about related events in the `LocationDetailsCard`.

## Screenshots
bild
bild

## Changes
* Adds model method to get events that are happening during a specific timespan
* Changes the map markers to be rendered as components instead of images.
* Adds number on top of marker icon. 
* Adds list of events in the `LocationDetailsCard`.

## Related issues
Resolves #1178

### CODE REVIEW

File: src/features/events/utils/eventUtils.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import theme from 'theme';
+
+export function getParticipantsStatusColor(
```
Comment (richardolsson) [inline src/features/events/utils/eventUtils.ts:3]: I like this, but it's a copy from the logic in other places. That's fine though, we will just have to refactor it later.

File: src/utils/dateUtils.ts [hunk 2]
```diff
@@ -74,6 +74,11 @@ export function isValidDate(date: Date): boolean {
     return false;
   }
 }
-export function dateIsBefore(first: Date, second: Date): boolean {
+
+export function isAfter(first: Date, second: Date): boolean {
```
Comment (richardolsson) [inline src/utils/dateUtils.ts:78]: Why rename this? I liked how it was called `dateIsBefore` because it made it clear(er) that it only looks at the date, not the time.

File: src/features/events/models/EventsModel.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    super();
+    this._orgId = orgId;
+    this._repo = new EventsRepo(env);
+  }
+
+  getAllEvents(): IFuture<ZetkinEvent[]> {
+    return this._repo.getAllEvents(this._orgId);
+  }
+
+  getEventsInTimespan(
```
Comment (richardolsson) [inline src/features/events/models/EventsModel.ts:27]: This function does exactly what it needs to do as far as I can tell, but the name implies something else to me.

The name implies to me that it find events that take place no earlier than `startString` and no later than `endString`. But what it really does is find events take place on or after the _date_ of `startString` (ignoring time) and on or before the _date_ of `endString` (ignoring time).

Again, this is correct, but the name implies something else to me.

File: src/features/events/l10n/messageIds.ts [hunk 4]
```diff
@@ -54,6 +54,8 @@ export default makeMessages('feat.events', {
     move: m('Move'),
     moveInstructions: m('Drag the pin to select location.'),
     noDescription: m('No description'),
+    noRelatedEvents: m('There are no other events in this location.'),
+    relatedEvents: m('Other events in this location'),
```
Comment (richardolsson) [inline src/features/events/l10n/messageIds.ts:58]: Maybe we should rephrase this as _Other events around this time_? It's obvious from the visual context that it relates to a certain location, is it not?

File: src/features/events/l10n/messageIds.ts [hunk 5]
```diff
@@ -54,6 +54,8 @@ export default makeMessages('feat.events', {
     move: m('Move'),
     moveInstructions: m('Drag the pin to select location.'),
     noDescription: m('No description'),
+    noRelatedEvents: m('There are no other events in this location.'),
```
Comment (richardolsson) [inline src/features/events/l10n/messageIds.ts:57]: I think we should clarify here that we mean _parallel_ events, i.e. events that happen at the same time as the event we're editing. How could we do that?

File: src/features/events/components/LocationModal/RelatedEventCard.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import { getParticipantsStatusColor } from 'features/events/utils/eventUtils';
+import { ZetkinEvent } from 'utils/types/zetkin';
+import ZUINumberChip from 'zui/ZUINumberChip';
+import ZUITimeSpan from 'zui/ZUITimeSpan';
+
+interface RelatedEventCardProps {
+  event: ZetkinEvent;
+}
+
+const RelatedEventCard: FC<RelatedEventCardProps> = ({ event }) => {
```
Comment (richardolsson) [inline src/features/events/components/LocationModal/RelatedEventCard.tsx:14]: Love this!

File: src/features/events/components/LocationModal/Map.tsx [hunk 7]
```diff
@@ -130,7 +133,22 @@ const Map: FC<MapProps> = ({
...
+                        ? divIcon({
+                            className: '',
+                            html: renderToStaticMarkup(<SelectedMarker />),
+                          })
+                        : divIcon({
+                            className: '',
+                            html: renderToStaticMarkup(
+                              <BasicMarker
+                                color={theme.palette.primary.main}
+                                events={noOfRelevantEvents}
```
Comment (richardolsson) [inline src/features/events/components/LocationModal/Map.tsx:147]: So nice, this! 💯

File: src/features/events/components/LocationModal/BasicMarker.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    width: 28,
+  },
+}));
+
+interface BasicMarkerProps {
+  color: string;
+  events: number;
+}
+
+const BasicMarker: FC<BasicMarkerProps> = ({ color, events }) => {
```
Comment (richardolsson) [inline src/features/events/components/LocationModal/BasicMarker.tsx:30]: Love this!

**Review by richardolsson — CHANGES_REQUESTED** [review]
I love how this works, but I have some minor comments on the code (naming etc) and want to raise some concerns about the text. I made some suggestions for new labels, but I would love your input on the text @ziggabyte because you're usually good with that kind of thing!

File: src/utils/dateUtils.ts (hunk 2 above)
Comment (ziggabyte) [reply src/utils/dateUtils.ts:78]: oops I think this was a mistake that I commited this, I sometimes rename functions/variables while working because the names confuse me personally, and change them back before committing heh

File: src/features/events/models/EventsModel.ts (hunk 3 above)
Comment (ziggabyte) [reply src/features/events/models/EventsModel.ts:27]: getParallelEvents ?

File: src/features/events/l10n/messageIds.ts (hunk 4 above)
Comment (ziggabyte) [reply src/features/events/l10n/messageIds.ts:58]: I had the same thoughts, that it is not completely clear that it means that they occur simultaneously. Like, it would be easy for me to assum they were just all events at that place no matter the time. I agree with your suggestion!

File: src/features/events/l10n/messageIds.ts (hunk 5 above)
Comment (ziggabyte) [reply src/features/events/l10n/messageIds.ts:57]: "There are no other events here at the same time." maybe?

**Review by richardolsson — CHANGES_REQUESTED** [review]
This is great! But there seems to have been a little mistake in 6af67f9a3bed6e07216172dad0441c9d10aae083 that needs to be rectified. Let me know if you want to double check the correct naming. Once that's been fixed I think we should merge!

**Review by richardolsson — CHANGES_REQUESTED** [review]
I'm so sorry about this…

File: src/utils/dateUtils.ts [hunk 9]
```diff
@@ -74,6 +74,11 @@ export function isValidDate(date: Date): boolean {
     return false;
   }
 }
+
+export function isAfter(first: Date, second: Date): boolean {
```
[the remark this answers is not in the document]
Comment (ziggabyte) [reply src/utils/dateUtils.ts:78]: completely agree! :D

**Review by richardolsson — APPROVED** [review]
Love it! Let's get this rose merged! 🌹
